src/Stratis.Bitcoin.Features.Wallet/WalletRPCController.cs

```diff
@@ -221,6 +225,46 @@ public GetTransactionModel GetTransaction(string txid)
             return model;
         }
 
+        [ActionName("listunspent")]
+        [ActionDescription("Returns an array of unspent transaction outputs belonging to this wallet.")]
+        public UnspentCoinModel[] ListUnspent(int minConfirmations = 1, int maxConfirmations = 9999999, string addressesJson = null)
+        {
+            List<BitcoinAddress> addresses = new List<BitcoinAddress>();
+            if (!string.IsNullOrEmpty(addressesJson))
+            {
+                JsonConvert.DeserializeObject<List<string>>(addressesJson).ForEach(i => addresses.Add(BitcoinAddress.Create(i, this.fullNode.Network)));
+            }
+            var accountReference = this.GetAccount();
+            IEnumerable<UnspentOutputReference> spendableTransactions = this.walletManager.GetSpendableTransactionsInAccount(accountReference, minConfirmations);
+
+            var unspentCoins = new List<UnspentCoinModel>();
+            foreach (var spendableTx in spendableTransactions)
+            {
+                int confirmations = spendableTx.Transaction.BlockHeight == null ? 0 : this.chain.Tip.Height - spendableTx.Transaction.BlockHeight.Value;
+                if (confirmations <= maxConfirmations)
+                {
+                    if (!addresses.Any() || addresses.Contains(BitcoinAddress.Create(spendableTx.Address.Address, this.fullNode.Network)))
+                    {
+                        unspentCoins.Add(new UnspentCoinModel()
+                        {
+                            Account = accountReference.AccountName,
+                            Address = spendableTx.Address.Address,
+                            Id = spendableTx.Transaction.Id,
+                            Index = spendableTx.Transaction.Index,
+                            Amount = spendableTx.Transaction.Amount,
+                            ScriptPubKeyHex = spendableTx.Transaction.ScriptPubKey.ToHex(),
+                            RedeemScriptHex = spendableTx.Transaction.ScriptPubKey.PaymentScript?.ToHex(), // Is this right?
```

Comment:
I don't think it is possible to trivially assign this like you are doing here (apart from the PaymentScript not being the same thing as the RedeemScript). It would require special handling within the wallet to support arbitrary P2SH payments, like we have partially done with the multisig wallet in the sidechain project. The crux of it is, the redeem script is not stored in the transaction data anywhere and cannot be derived from it. It has to be maintained somewhere else at the time of transaction creation, or communicated to the party wanting to spend the transaction out-of-band.

Reply:
Kevin is right here.
Our wallet only generates P2PKH addresses and not P2SH ones. As such, we don't store or even have redeem scripts in our outputs. You can give this a null and maybe add a TOD to say that this needs to be reviewed when we support P2SH addresses, whether it's multisig or arbitrary ones.